core/src/main/java/org/apache/accumulo/core/util/threads/ThreadPools.java

```diff
@@ -65,8 +66,18 @@ public ExecutionError(String message, Throwable cause) {
   // the number of seconds before we allow a thread to terminate with non-use.
   public static final long DEFAULT_TIMEOUT_MILLISECS = 180000L;
 
+  private static final ThreadPools SERVER_INSTANCE = new ThreadPools(Threads.UEH);
+
+  public static final ThreadPools getServerThreadPools() {
+    return SERVER_INSTANCE;
+  }
```

Comment:
I don't know if it's easy to do in all cases, but I was thinking that instead of this, it can share the same sort of code as for the client-side cases.

Essentially, ClientContext would have a `.threadPools()` method that returns an instance of `ThreadPools` constructed using the UEH provided when AccumuloClient was built (default if not provided). When ServerContext is constructed, it just passes the built-in UEH for the server-side code to ClientContext's constructor in `super()`.

So, instead of calling `ThreadPools.getServerThreadPools().whatever`, you'd just use `context.threadPools().whatever` everywhere, regardless of client or server context. That way, there's no static instance of ThreadPools anywhere, but everything still works as expected. (`context.threadPools()` is also shorter than `ThreadPools.getServerThreadPools()`)

Reply:
I don't know that in all cases server-side that the ServerContext is available to the code. I would have to wire it up in all cases where it's not reachable.

Reply:
To be clear, only ClientContext would be necessary to wire up. ServerContext isn't available in the core jar. However, it is a subclass of ClientContext, so it would still work if only a ClientContext. But even then, ClientContext might not be available everywhere, and might need to be wired up... that's one of the things we've been fighting against to avoid static state in the JVM and the problems that come with it, so it's not a new issue. But, I was just hoping we could try to avoid a few additional occurrences of it that we'd have to fix later.

In any case, that could be treated as a follow-on. I think the main things to address before this is merged in is to move the static thread pools that were converted to instance members for cleanup into ClientContext, so they exist in a Closeable object and don't need a separate Cleaner, and so they can be reused for the cleanup tasks they perform across multiple instances of the objects they clean up.

Reply:
Won't they still need a Cleaner in the case where the client is not closed?

Reply:
I thought the client context already has a cleaner that ensures it is closed, because of the legacy Instance/Connector stuff that may not clean up the client context.

However, even if it doesn't, Cleaner's (replacing finalizers) are a hack to close things during garbage collection, and are primarily useful for objects that don't explicitly get closed. For objects whose lifecycle is user-facing, like our Closeable clients, we don't need to specify a Cleaner... we (or the user) can just close them. Java helps us with this by giving us warnings about resource leaks and potential resource leaks for unclosed Closeable objects.

Reply:
I see the something in SingletonReservation that closes AccumuloClients if that is what you are referencing. I will remove the Cleaners from ClientContext.

Reply:
It is. Thanks.